### PR TITLE
Add commit status on repo and user pull request lists

### PR DESCRIPTION
**Pull request by lunny**
And something confusing me is should I use `issue.PullRequest.MergeBase` as the last commit of a pull request? Anyway, see the screenshots below:

### CODE REVIEW

**Comment by lafriks** [comment]
Integration test would be nice for this

**Comment by lunny** [comment]
@lafriks It seems some difficult since our integration tests didn't run a drone instance?

**Comment by lafriks** [comment]
@lunny you can just call status api to set status for commit you dont need real CI

File: routers/user/home.go
Comment (Morlinest) [inline routers/user/home.go]: Can you extract this block to new function? I've found this (almost) same code block 3 times in existing codebase.

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: routers/user/home.go
Comment (lunny) [reply routers/user/home.go]: I think they are different `for ... range`.

Comment (Morlinest) [reply routers/user/home.go]: It's same, the only difference is you are adding result of `models.CalcCommitStatus(statuses)` to slice. Instead of calling 2 functions and 1 err check, you can call just 1 function (err check can be inside too, it changes nothing). Best thing would be to rename `GetLatestCommitStatus` to `GetCommitStatuses` (as it returns array, not just 1 status) and create/use `GetLatestCommitStatus` to return just one status (for now, this few lines of code can be used, but I want to change it to pure SQL select).

File: templates/repo/issue/list.tmpl
Comment (bkcsoft) [inline templates/repo/issue/list.tmpl]: indentation

**Comment by lunny** [comment]
In fact I prefer to implement `GetLatestCommitStatusByIDs`

File: routers/user/home.go
Comment (Morlinest) [reply routers/user/home.go]: ok, `GetCommitStatuses` is here also, I don't understand `GetLatestCommitStatus` implementation... As I understand, it should return only 1 status in slice.

**Comment by bkcsoft** [comment]
@lunny `ByIDs` ? What ID?

**Comment by lunny** [comment]
@bkcsoft `GetLatestCommitStatusByRepoIDsAndCommitIDs(repoIDs, commitIDs)`

File: routers/repo/issue.go
Comment (ethantkoenig) [inline routers/repo/issue.go]: Let's not ignore the error returned here

File: routers/user/home.go
Comment (ethantkoenig) [inline routers/user/home.go]: Let's not ignore the error returned here

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
@lunny That function name is _waaaay_ to long 😂 I say that `(repoID, commitID)` is enough documentation to just call it `GetLatestCommitStatuses` ;)

File: routers/repo/issue.go
Comment (lunny) [reply routers/repo/issue.go]: done.

File: routers/user/home.go
Comment (lunny) [reply routers/user/home.go]: done.

File: templates/repo/issue/list.tmpl
Comment (lunny) [reply templates/repo/issue/list.tmpl]: done.

File: models/status.go
Comment (Morlinest) [inline models/status.go]: Is not this select everything you need to get latest commit for each repo commit (I was just writing about adding columns to "group by" when you updated it)? You can use this as subselect (just remove `repo_id` from select) for mainy "where" clause.

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@lafriks integration tests added

**Comment by lunny** [comment]
@Morlinest I think the best way is add a field `LastCommitID` on `PullRequest` and fill in it when push commit to the headbranch. So I will do it on another PR and keep this one simple.

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: routers/repo/issue.go
Comment (lunny) [reply routers/repo/issue.go]: Yes, you are right. But I think store the last commit id in pullrequest table is the best way. I will send another PR to do that.

**Comment by Morlinest** [comment]
@lunny OK, I'll wait for new PR and updated version because I have more comments to this PR :)

File: routers/repo/issue.go
Comment (bkcsoft) [reply routers/repo/issue.go]: If you do, call it `RepoSHA` instead of `CommitOption` since that better explains the purpose of the struct. Though in general I dislike `FunctionOptions` structs since they hide the arguments from the docs.

**Comment by lunny** [comment]
@Morlinest ~I will update this after https://github.com/go-gitea/gitea/pull/2548 merged.~

File: routers/repo/issue.go
Comment (Morlinest) [reply routers/repo/issue.go]: @bkcsoft But 2 slices is not solution to that as you can't be sure about the order nor dependency between values on same key. I don't think that you are hiding something with structs in go.

File: integrations/editor_test.go
Comment (ethantkoenig) [inline integrations/editor_test.go]: nit: you can use `GetCSRF(..)`, so you don't have to manually make GET request just for the csrf token.

File: templates/repo/commit_status.tmpl
Comment (ethantkoenig) [inline templates/repo/commit_status.tmpl:1]: This file should use tabs instead of spaces

File: routers/user/home.go
Comment (ethantkoenig) [inline routers/user/home.go]: nit: could we call the variable `gitRepo`, or something more descriptive? Without context, it's not obvious (at least not to me) that rep stands for repo.

File: templates/repo/issue/list.tmpl
Comment (ethantkoenig) [inline templates/repo/issue/list.tmpl]: Is `$index` ever used? Similarly for the other template.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: templates/repo/issue/list.tmpl
Comment (lunny) [reply templates/repo/issue/list.tmpl]: Seems no needed after my second push.

**Comment by lunny** [comment]
will resolve #996

**Comment by lunny** [comment]
@ethantkoenig @Morlinest @bkcsoft all done. Please review and I think it's ready for review and no need wait for #2548. Of course, we can improve the performance on #2548 or after #2548 merged.

File: models/status.go
Comment (ethantkoenig) [inline models/status.go]: Suggestion: change the type of this function's argument to be a slice of issues instead of a slice of these anonymous structs.

I see two main benefits:
1. The code will be cleaner, since we won't need to use this anonymous struct in several different places.
2. The code surrounding the calls to this function inside `repo.Issues(..)` and `home.Issues(..)` are nearly identical. We can move that shared code inside this function by changing the argument type.

File: integrations/repo_pull_status_test.go
Comment (ethantkoenig) [inline integrations/repo_pull_status_test.go]: Please remove

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by lafriks** [comment]
@lunny can you resolve conflicts, I would love to see this in 1.4.0 :)

File: models/pull.go
Comment (bkcsoft) [inline models/pull.go]: Why not pull this from the Repo?

File: models/status.go
Comment (bkcsoft) [inline models/status.go:162]: Why not use `GetLatestCommitStatus` ?

If you need to fetch from multiple SHAs at the same time, just make a `GetLatestCommitsStatus(repo *Repository, shas []string, page int)` and fetch the `shas` before that 🤔

Comment (bkcsoft) [reply models/status.go:162]: Also, I would return a `map[string][]*CommitStatus` to that you can `ret[shas[0]` when using them.

Comment (bkcsoft) [reply models/status.go:162]: OR, you can have a `func (*Issue) PopulateLatestCommitStatus()` that fetches them per Issue 🤔

File: routers/repo/issue.go
Comment (bkcsoft) [inline routers/repo/issue.go:197]: Why this change?

**Review by bkcsoft — CHANGES_REQUESTED** [review]
Other than these I don't see any reason to hold this up any longer :)

**Comment by appleboy** [comment]
@lunny build fail.

**Comment by bastienlemaitre** [comment]
Any news ?

**Review by appleboy — APPROVED** [review]

**Comment by bastienlemaitre** [comment]
@lunny ? @bkcsoft ?

**Comment by lunny** [comment]
@appleboy but mysql-test failed. It's strange.

**Comment by appleboy** [comment]
@lunny conflict.

**Comment by lunny** [comment]
@appleboy done. But I haven't resolve the test failed issue.

**Comment by jonasfranz** [comment]
@lunny conflicted

**Comment by strk** [comment]
@lunny are you still planning to work on this ?

**Comment by lunny** [comment]
@strk The tests always failed and I don't know why. If the tests PASS, I think it's enough to review. Could you help me with this PR?

File: models/pull.go
Comment (vijfhoek) [inline models/pull.go]: Not sure about the styleguide here, but shouldn't the tags be aligned, and why 3 spaces between `LastCommitID` and `string`?

**Review by vijfhoek — CHANGES_REQUESTED** [review]

**Comment by vijfhoek** [comment]
Is there anything I can do to help moving this forward a bit quicker? This PR has been open for far too long, and I really need this feature.

**Comment by lunny** [comment]
@SijmenSchoon if you want to work on this, just take it and create another PR. I think this PR needs to fix tests. I'm now busy on other PRs.

**Comment by lunny** [comment]
Since #6465 merged, I will close this one.

**Pull request closed without merging**
